### PR TITLE
feat: update ci/cd to lint and typecheck

### PR DESCRIPTION
**Pull request by karrui**
This PR introduces significant refinements and restructuring to the GitHub Actions workflows to enhance the CI processes. It includes renaming the workflow file from `main.yml` to `ci.yml`, adding new jobs for linting, formatting, type-checking, and testing, as well as uninstalling some dependencies and addressing various minor issues in code formatting and types.

### CODE REVIEW

**Comment by karrui** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/opengovsg/isomer/284?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#284** <a href="https://app.graphite.dev/github/pr/opengovsg/isomer/284?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#282** <a href="https://app.graphite.dev/github/pr/opengovsg/isomer/282?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#283** <a href="https://app.graphite.dev/github/pr/opengovsg/isomer/283?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#281** <a href="https://app.graphite.dev/github/pr/opengovsg/isomer/281?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#280** <a href="https://app.graphite.dev/github/pr/opengovsg/isomer/280?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#279** <a href="https://app.graphite.dev/github/pr/opengovsg/isomer/279?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @karrui and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by karrui** [comment]
This one i WIP first, might have a shitton of errors whilst i fix all the actions lol

**Review by seaerchin — APPROVED** [review]
just a qq here - since we omit `npm run build`, i think our assumption is that passing `typecheck + lint` equals to passing `build`. 

i don't foresee any problems with that assumption but just wanted to ask if this was necessarily true. going off what i can see for next, this is but idk fi they do anything under the hood

File: package.json [hunk 1]
```diff
@@ -11,7 +11,12 @@
     "storybook": "dotenv -- turbo run storybook --parallel",
     "storybook:build": "turbo run storybook:build --parallel",
     "teardown": "turbo teardown",
-    "test-start": "turbo test-start"
+    "test-start": "turbo test-start",
+    "typecheck": "turbo typecheck",
+    "format": "turbo format --continue -- --cache --cache-location node_modules/.cache/.prettiercache --ignore-path ./.prettierignore",
+    "format:fix": "turbo format --continue -- --write --cache --cache-location node_modules/.cache/.prettiercache  --ignore-path ./.prettierignore",
+    "lint": "turbo lint --continue -- --cache --cache-location node_modules/.cache/.eslintcache",
+    "lint:fix": "turbo lint --continue -- --fix --cache --cache-location node_modules/.cache/.eslintcache"
```
[the remark this answers is not in the document]
Comment (karrui) [reply package.json:19]: ya for now use it like this first, fix when break lol

**Comment by karrui** [comment]
> just a qq here - since we omit `npm run build`, i think our assumption is that passing `typecheck + lint` equals to passing `build`.
> 
> i don't foresee any problems with that assumption but just wanted to ask if this was necessarily true. going off what i can see for next, this is but idk fi they do anything under the hood

Got test also, that one needs build

**Comment by karrui** [comment]
### Merge activity

* **Jul 9, 6:24 AM EDT**: @karrui started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/opengovsg/isomer/284).

**Pull request merged**